bin/peggy-cli.js

```diff
@@ -610,6 +750,56 @@ class PeggyCLI extends Command {
     });
   }
 
+  /**
+   * @param {import("../lib/peg.js").ast.Grammar} ast
+   * @returns {Promise<void>}
+   */
+  async writeDTS(ast) {
+    if (!this.dtsFile) {
+      return;
+    }
+    let template = await fs.promises.readFile(
+      path.join(__dirname, "generated_template.d.ts"), "utf8"
+    );
+    let startRules = (this.argv.allowedStartRules || [ast.rules[0].name]);
+    if (startRules.includes("*")) {
+      startRules = ast.rules.map(r => r.name);
+    }
+    const qsr = startRules.map(r => `"${r}"`);
+
+    template = template.replace("$$$StartRules$$$", qsr.join(" | "));
+    template = template.replace("$$$DefaultStartRule$$$", qsr[0]);
+
+    const out = fs.createWriteStream(this.dtsFile);
+    out.write(template);
+
+    const types = /** @type {Record<string, string>|undefined} */(
+      this.progOptions.returnTypes
+    ) || {};
+    for (const sr of startRules) {
+      out.write(`
+export function ParseFunction<Options extends ParseOptions<"${sr}">>(
+  input: string,
+  options?: Options,
+): ${types[sr] || "any"};
```

Comment:
It doesn't seem like `ParseFunction` is actually exported. In my case it's:
```js
    return {
      StartRules: ["Expression"],
      SyntaxError: peg$SyntaxError,
      parse: peg$parse
    };
```

After bundling.

I can confirm that nothing else shows up at runtime either.
![image](https://github.com/user-attachments/assets/9338d153-ca95-43af-a490-8c343587ff32)

Reply:
You're supposed to get one or more overloads of ParseFunction, one for each allowedStartRule.  I wonder what you're doing that those aren't getting added.  Can you try a simple example with a basic grammar like `foo = '1'` and see what happens?

You'll want some `returnTypes`, which I'll document before this lands.

Reply:
Hey! The project I'm working with isn't actually my own, I am working on writing `.d.ts` files for it though. It's for a project called FoundryVTT which has 
this [grammar](https://github.com/user-attachments/files/16400263/grammar.txt) (file renamed to grammar.txt so GitHub will let me upload it).

They're using `"peggy": "^4.0.2",` so it's not really that old. I tried running peggy on the example `foo = '1'` grammar you provided and I still didn't get `ParseFunction`. I ran the command as `./node_modules/peggy/bin/peggy.js foo.pegjs -o foo` maybe that has something to do with this?

I can provide a full minimal reproduction if you want.

Reply:
If you can create a minimal repro, that would be very helpful.  I don't want to merge this PR until we've figured this out.

Can you try my minimal version at https://gist.github.com/hildjj/67c0b19b0dae7f4f88b040fe553a7a4f please?

Reply:
I downloaded your minimal version. Unfortunately I was still unable to get an exported `ParseFunction`. Perhaps this is user error?

Here's my terminal session:
![demo](https://github.com/user-attachments/assets/72367096-c1f4-41e3-8ae8-270d758d92d9)

If you want to take a closer look here's my [final result](https://github.com/user-attachments/files/16403031/peggy.repro.zip) (node_modules included).

I'm on WSL2, specifically:
```
Linux 5.15.153.1-microsoft-standard-WSL2 #1 SMP Fri Mar 29 23:14:13 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

I can send more system information if you want.

Reply:
I searched the rest of the repository for `ParseFunction` and couldn't find it outside of this PR which only adds it to the `.d.ts` files.

This makes me think that maybe I confused you with the way I initially phrased it? I _am_ seeing `ParseFunction` in the `.d.ts` files but not in the `.js` files. This is a problem because `ParseFunction` is exported in the `.d.ts` files but won't exist at runtime.

This means I could write:
```ts
import { ParseFunction } from "foo.js"
```

And TS would be perfectly happy compiling this because when it sees `foo.js` it'll also look for `foo.d.ts` and in this case it'll find it and use it as the soruce of truth. Because there's a line saying `export function ParseFunction(...)` in the generated `.d.ts` file it'll then allow me to import `ParseFunction` but it'll then error at runtime.

This is what I meant by it not being "actually exported"-by saying `export function ParseFunction` you're saying it should be exported in the `.d.ts` file but when I looked at the corresponding `.js` file it's not actually so it'll be a runtime error if someone tries instead of a compile time error.

If the only reason why `ParseFunction` is there is as an implementation detail you should replace `export function` with `declare function` and that should fix everything. The `declare` means tsc won't whine about a missing function body but it also won't make it accessible to the outside world.

Reply:
See if this latest change fixes it for you?

I'm interested in any other TypeScript insight you might have.  I'm particularly bad at creating .d.ts files by hand.